src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1382 @@
+            auto normalized = resolved.clone(Url::EquivalenceOption::NORMALIZE_PATH);
+            return js.str(normalized.getHref());
+          } else {
+            // If the specifier could not be parsed and resolved successfully,
+            // the spec says to return null.
+            return js.null();
+          }
+        });
+
+        if (meta->CreateDataProperty(js.v8Context(),
+            v8::Local<v8::String>(js.strIntern("resolve"_kj)),
+            resolve).IsNothing()) {
+          return;
+        }
+      }
+    }, [&](Value exception) {
+      // It would be exceedingly odd to end up here but we handle it anyway,
+      // just to ensure that we do not crash the isolate. The only thing we'll
+      // do is rethrow the error.
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+    });
+  } catch(...) {
+    kj::throwFatalException(kj::getCaughtExceptionAsKj());
+  }
+}
+
+// The callback v8 calls when dynamic import(...) is used.
+v8::MaybeLocal<v8::Promise> dynamicImport(
+    v8::Local<v8::Context> context,
+    v8::Local<v8::Data> host_defined_options,
+    v8::Local<v8::Value> resource_name,
+    v8::Local<v8::String> specifier,
+    v8::Local<v8::FixedArray> import_assertions) {
+  auto isolate = context->GetIsolate();
+  auto& js = Lock::from(isolate);
+  try {
+    return js.tryCatch([&]() -> v8::MaybeLocal<v8::Promise> {
+      auto spec = js.toString(specifier);
+
+      // The proposed specification for import assertions strongly recommends that
+      // embedders reject import attributes and types they do not understand/implement.
+      // This is because import attributes can alter the interpretation of a module.
+      // Throwing an error for things we do not understand is the safest thing to do
+      // for backwards compatibility.
+      //
+      // For now, we do not support any import attributes, so if there are any at all
+      // we will reject the import.
+      if (!import_assertions.IsEmpty()) {
+        if (import_assertions->Length() > 0) {
+          auto error = js.typeError("Import attributes are not supported");
+          v8::Local<v8::Promise::Resolver> resolver;
+          if (!v8::Promise::Resolver::New(context).ToLocal(&resolver) ||
+              resolver->Reject(context, error).IsNothing()) {
+            return v8::MaybeLocal<v8::Promise>();
+          }
+          return resolver->GetPromise();
+        };
+      }
+
+      Url referrer = ([&] {
+        if (resource_name.IsEmpty()) {
+          return ModuleBundle::BundleBuilder::BASE.clone();
+        }
+        auto str = js.toString(resource_name);
+        return KJ_ASSERT_NONNULL(Url::tryParse(str.asPtr()));
+      })();
+
+      KJ_IF_SOME(url, referrer.tryResolve(spec.asPtr())) {
+        auto normalized = url.clone(Url::EquivalenceOption::NORMALIZE_PATH);
+        auto& registry = IsolateModuleRegistry::from(isolate);
+        return registry.dynamicResolve(js, kj::mv(normalized), kj::mv(referrer), spec);
```

Comment:
Note for reviewers: As discussed in the overview call, the new module registry is *currently* implemented such that dynamic imports are resolved and evaluated within the active request context (`IoContext`) if there is one. This is where that happens. If we decide that dynamic imports should NOT be aware of the request context, then we should change this here to ensure that we pop out of the context, perform the evaluation, then come back into it the way we currently do with the old registry.